### PR TITLE
Simplify the integration test for node lifecycle manager

### LINKED ISSUE #111183: Simplify the "TestTaintBasedEvictions" integration test in lifecycle_test.go
The following things should be simplified:
- use "NodeDisruptionExclusion" feature instead of  periodic heartbeat events to simplify the logic, it was also requested in this comment:
```
			// Regularly send heartbeat event to APIServer so that the cluster doesn't enter fullyDisruption mode.
			// TODO(Huang-Wei): use "NodeDisruptionExclusion" feature to simply the below logic when it's beta.
```
- eliminate scheduler from this test

### PR DESCRIPTION
**Pull request by mimowo**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Simplifies the setup in the integration test for node lifecycle manager. In particular:
- uses "NodeDisruptionExclusion" feature instead of  periodic heartbeat events to simplify the logic, it was also requested in this comment:
```
			// Regularly send heartbeat event to APIServer so that the cluster doesn't enter fullyDisruption mode.
			// TODO(Huang-Wei): use "NodeDisruptionExclusion" feature to simply the below logic when it's beta.
```
- eliminates scheduler from this test
- replaces the custom code for awaiting for a condition with built-in wait.PollImmediate()

#### Which issue(s) this PR fixes:

Fixes #111183

#### Special notes for your reviewer:

None

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
NONE
```

### CODE REVIEW

File: test/integration/node/lifecycle_test.go [hunk 1]
```diff
@@ -297,168 +292,89 @@ func TestTaintBasedEvictions(t *testing.T) {
 			// Waiting for all controllers to sync
 			externalInformers.Start(testCtx.Ctx.Done())
 			externalInformers.WaitForCacheSync(testCtx.Ctx.Done())
-			testutils.SyncInformerFactory(testCtx)
```
Comment (alculquicondor) [inline test/integration/node/lifecycle_test.go:300]: are you sure this line is not necessary?

Comment (mimowo) [reply test/integration/node/lifecycle_test.go:300]: Yes, actually we cannot even call it in the current setup. because the function calls `Start` and `WaitForCacheSync` on `testCtx.InformerFactory` which is nil in this test (after getting rid of scheduler). 
However, using this function could replace the two lines above `externalInformers.Start(testCtx.Ctx.Done())` and `externalInformers.WaitForCacheSync(testCtx.Ctx.Done())`  if we earlier assigned `testCtx.InformerFactory = externalInformers`, but it isn't a clear gain to me. The function `SyncInformerFactory` seems to only be used by tests for Scheduler currently.

File: test/integration/node/lifecycle_test.go [hunk 2]
```diff
@@ -132,7 +132,7 @@ func TestEvictionForNoExecuteTaintAddedByUser(t *testing.T) {
 				true,             // Run taint manager
 			)
 			if err != nil {
-				t.Errorf("Failed to create node controller: %v", err)
+				t.Fatalf("Failed to create node controller: %v", err)
 				return
```
Comment (alculquicondor) [inline test/integration/node/lifecycle_test.go:136]: no need for return anymore

File: test/integration/node/lifecycle_test.go [hunk 3]
```diff
@@ -285,7 +285,7 @@ func TestTaintBasedEvictions(t *testing.T) {
...
```
Comment (alculquicondor) [inline test/integration/node/lifecycle_test.go:289]: here too.

**Review by alculquicondor — COMMENTED** [review]
/lgtm

As far as I can tell, this PR is not reducing flakiness or preventing goroutine leaks, so it can wait for 1.26

File: test/integration/node/lifecycle_test.go (hunk 3 above)
Comment (mimowo) [reply test/integration/node/lifecycle_test.go]: Indeed, updated ans squashed

File: test/integration/node/lifecycle_test.go (hunk 2 above)
Comment (mimowo) [reply test/integration/node/lifecycle_test.go]: Indeed, updated and squashed

**Comment by mimowo** [comment]
@alculquicondor removed the unnecessary returns and squashed the commits. 
Please renew the LGTM if no futher remarks.

**Comment by mimowo** [comment]
/assign @random-liu
1.25.0 is released so it seems it is ok to merge this PR

**Comment by haircommander** [comment]
/cc @Huang-Wei

as the original TODO writer

changes LGTM, besides one nit that wasn't clear to me (though I am not familiar with the node lifecycle manager)

File: test/integration/node/lifecycle_test.go [hunk 4]
```diff
@@ -290,175 +284,95 @@ func TestTaintBasedEvictions(t *testing.T) {
...
+					if test.tolerationSeconds == 0 {
 						return pod.DeletionTimestamp != nil, nil
 					}
 					if seconds, err := testutils.GetTolerationSeconds(pod.Spec.Tolerations); err == nil {
-						return seconds == tolerationSeconds[i], nil
+						return seconds == test.tolerationSeconds, nil
 					}
 					return false, nil
-				}, t)
+
```
Comment (Huang-Wei) [inline test/integration/node/lifecycle_test.go:367]: nit: remove this empty line.

**Comment by Huang-Wei** [comment]
Thanks @mimowo. Changes LGTM (but I don't have the approve privilege).

File: test/integration/node/lifecycle_test.go [hunk 5]
```diff
@@ -290,175 +284,95 @@ func TestTaintBasedEvictions(t *testing.T) {
...
 			go nc.Run(testCtx.Ctx)
-			go testCtx.Scheduler.Run(testCtx.Ctx)
 
 			nodeRes := v1.ResourceList{
 				v1.ResourceCPU:    resource.MustParse("4000m"),
 				v1.ResourceMemory: resource.MustParse("16Gi"),
 				v1.ResourcePods:   resource.MustParse("110"),
 			}
 
+			nodeIndex := 1
```
[the remark this answers is not in the document]
Comment (Huang-Wei) [reply test/integration/node/lifecycle_test.go:303]: yup, it's randomly picked.

File: test/integration/node/lifecycle_test.go (hunk 4 above)
Comment (mimowo) [reply test/integration/node/lifecycle_test.go]: Done

File: test/integration/node/lifecycle_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (mimowo) [reply test/integration/node/lifecycle_test.go]: added the suggested comment

**Comment by mimowo** [comment]
/assign @dchen1107
@haircommander @Huang-Wei the suggested changes are applied and squashed.

File: test/integration/node/lifecycle_test.go [hunk 6]
```diff
@@ -152,7 +145,7 @@ func TestEvictionForNoExecuteTaintAddedByUser(t *testing.T) {
 			for index := range nodes {
 				nodes[index], err = cs.CoreV1().Nodes().Create(testCtx.Ctx, nodes[index], metav1.CreateOptions{})
 				if err != nil {
-					t.Errorf("Failed to create node, err: %v", err)
+					t.Fatalf("Failed to create node, err: %v", err)
```
Comment (aojea) [inline test/integration/node/lifecycle_test.go:148]: it's better if we do not conflate unrelated changes on the same commit

File: test/integration/node/lifecycle_test.go [hunk 7]
```diff
@@ -206,19 +198,20 @@ func TestTaintBasedEvictions(t *testing.T) {
 			TerminationGracePeriodSeconds: &gracePeriod,
 		},
 	}
-	tolerationSeconds := []int64{200, 300, 0}
```
Comment (aojea) [inline test/integration/node/lifecycle_test.go:209]: yeah, this is weird, why wasn't part of the struct before?

File: test/integration/node/lifecycle_test.go [hunk 8]
```diff
@@ -290,175 +284,94 @@ func TestTaintBasedEvictions(t *testing.T) {
...
 			go nc.Run(testCtx.Ctx)
-			go testCtx.Scheduler.Run(testCtx.Ctx)
 
 			nodeRes := v1.ResourceList{
 				v1.ResourceCPU:    resource.MustParse("4000m"),
 				v1.ResourceMemory: resource.MustParse("16Gi"),
 				v1.ResourcePods:   resource.MustParse("110"),
 			}
 
+			nodeIndex := 1 // the exact node doesn't matter, pick one
```
Comment (aojea) [inline test/integration/node/lifecycle_test.go:303]: this should be better placed with `nodeCount`  , since those are related constant for this test , and we can avoid problems in the future if both go out of sync

File: test/integration/node/lifecycle_test.go [hunk 9]
```diff
@@ -290,175 +284,94 @@ func TestTaintBasedEvictions(t *testing.T) {
...
+					t.Fatalf("Failed to create node: %q, err: %v", klog.KObj(node), err)
 				}
 			}
 
-			neededNode := nodes[1]
 			if test.pod != nil {
+				test.pod.Spec.NodeName = nodes[nodeIndex].Name
 				test.pod.Name = fmt.Sprintf("testpod-%d", i)
 				if len(test.pod.Spec.Tolerations) > 0 {
-					test.pod.Spec.Tolerations[0].TolerationSeconds = &tolerationSeconds[i]
```
Comment (aojea) [inline test/integration/node/lifecycle_test.go:341]: do we need the `i` in line 253 now?

> 	for i, test := range tests {

there are too many `i` in this test :shrug:

**Comment by aojea** [comment]
> Thanks @mimowo. Changes LGTM (but I don't have the approve privilege).

/approve

Approving on behalf of @Huang-Wei , holding because I have some nits, but  feel free to unhold 
/hold

File: test/integration/node/lifecycle_test.go (hunk 6 above)
Comment (mimowo) [reply test/integration/node/lifecycle_test.go:148]: I think replacing the Errorf with Fatalf for setup makes the test simpler (which is the goal of the PR).

File: test/integration/node/lifecycle_test.go (hunk 7 above)
Comment (mimowo) [reply test/integration/node/lifecycle_test.go:209]: It seems for 'historical reasons'. I've traced it back to this commit https://github.com/kubernetes/kubernetes/commit/ca10e616bf88ee0f0ed00762c92948761dd56d62. It seems the test was first added as a non-table e2e test. Later it was converted to a table integration test, but the list was not refactored. Seems reasonable to do it now.

File: test/integration/node/lifecycle_test.go (hunk 8 above)
Comment (mimowo) [reply test/integration/node/lifecycle_test.go]: Moved up.

File: test/integration/node/lifecycle_test.go (hunk 7 above)
Comment (aojea) [reply test/integration/node/lifecycle_test.go:209]: agreed, it is much better this way

File: test/integration/node/lifecycle_test.go (hunk 6 above)
Comment (aojea) [reply test/integration/node/lifecycle_test.go:148]: yeah, but think that those are 2 different tests, if we somehow figure out after merge that we broke something, we have to revert the commit, and we want to revert only the code that break something,
In this case, you should have one commit per test, we somehow organize commits per topic to facilitate reviews, but keeping a clean git history helps the code maintenance , is not a big deal in this case, but hygiene always help

File: test/integration/node/lifecycle_test.go (hunk 9 above)
Comment (mimowo) [reply test/integration/node/lifecycle_test.go:341]: Agree, renamed the outer `i` as `testIndex`

File: test/integration/node/lifecycle_test.go (hunk 6 above)
Comment (mimowo) [reply test/integration/node/lifecycle_test.go:150]: I get the point about keeping commits isolated and clean. However, I think in this case
case it makes sense to apply the same changes to both tests in the file (`TestTaintBasedEvictions` and `TestEvictionForNoExecuteTaintAddedByUser`):
- move `nodeIndex` next to `nodeCount`
- add a comment to `nodeIndex`
- using Fatalf instead of Errorf 

Applying them to only one of the tests would make another reader puzzled it seems.

Comment (mimowo) [reply test/integration/node/lifecycle_test.go:150]: I've split the changes into two separate commits within the PR. Hope it makes sense.

Comment (aojea) [reply test/integration/node/lifecycle_test.go:150]: ah ok, then I misread it, sorry for the obvious comment then :smile:

**Comment by aojea** [comment]
> /lgtm

carrying over the lgtm, my comments and fixes were mostly aesthetic since the 2 previous lgtm

/lgtm

Thanks for your patience :smile:

File: test/integration/node/lifecycle_test.go (hunk 6 above)
Comment (mimowo) [reply test/integration/node/lifecycle_test.go:150]: no worries, thanks

**Comment by aojea** [comment]
/hold cancel

I forget about it, sorry

**Comment by aojea** [comment]
/hold

I run it with `stress` and it flakes :thinking: 

```
I0919 10:50:33.688690 3888769 etcd.go:107] starting etcd on http://127.0.0.1:40041
I0919 10:50:33.688830 3888769 etcd.go:113] storing etcd data in: /tmp/integration_test_etcd_data3544937043
{"level":"warn","ts":"2022-09-19T10:50:33.737+0200","caller":"auth/store.go:1220","msg":"simple token is not cryptographically signed"}
I0919 10:50:35.179153 3888769 serving.go:342] Generated self-signed cert (/tmp/test-integration-TestTaintBasedEvictions_Taint_based_evictions_for_NodeNotReady_and_200_tolerationseconds2974214670/apiserver.crt, /tmp/test-integration-TestTaintBasedEvictions_Taint_based_evictions_for_NodeNotReady_and_200_tolerationseconds2974214670/apiserver.key)
I0919 10:50:35.179173 3888769 server.go:554] external host was not specified, using 192.168.1.150
I0919 10:50:35.762568 3888769 shared_informer.go:273] Waiting for caches to sync for node_authorizer
I0919 10:50:35.763654 3888769 plugins.go:158] Loaded 7 mutating admission controller(s) successfully in the following order: NamespaceLifecycle,LimitRanger,DefaultTolerationSeconds,DefaultStorageClass,RuntimeClass,DefaultIngressClass,MutatingAdmissionWebhook.
I0919 10:50:35.763670 3888769 plugins.go:161] Loaded 9 validating admission controller(s) successfully in the following order: LimitRanger,PodSecurity,PersistentVolumeClaimResize,RuntimeClass,CertificateApproval,CertificateSigning,CertificateSubjectRestriction,ValidatingAdmissionWebhook,ResourceQuota.
I0919 10:50:35.764165 3888769 instance.go:261] Using reconciler: lease
I0919 10:50:35.880393 3888769 instance.go:574] API group "internal.apiserver.k8s.io" is not enabled, skipping.
W0919 10:50:36.059199 3888769 genericapiserver.go:656] Skipping API authentication.k8s.io/v1beta1 because it has no resources.
W0919 10:50:36.059219 3888769 genericapiserver.go:656] Skipping API authentication.k8s.io/v1alpha1 because it has no resources.
W0919 10:50:36.061639 3888769 genericapiserver.go:656] Skipping API authorization.k8s.io/v1beta1 because it has no resources.
W0919 10:50:36.067740 3888769 genericapiserver.go:6
…
2m10s: 96 runs so far, 17 failures (17.71%)
2m15s: 96 runs so far, 17 failures (17.71%)
2m20s: 96 runs so far, 17 failures (17.71%)
2m25s: 100 runs so far, 17 failures (17.00%)
2m30s: 108 runs so far, 17 failures (15.74%)
2m35s: 120 runs so far, 17 failures (14.17%)
2m40s: 120 runs so far, 17 failures (14.17%)
```

**Comment by mimowo** [comment]
> I run it with `stress` and it flakes 🤔
> 
Could you please test if this flakes for you on master?

**Comment by aojea** [comment]
> /hold

/hold cancel

same behavior in master , checking the logs I think that is related to the test setup and not using the shared etcd instance

```
/tmp/go-stress-20220919T112459-1844619485
I0919 11:26:40.267016 3902624 etcd.go:107] starting etcd on http://127.0.0.1:43411
I0919 11:26:40.267121 3902624 etcd.go:113] storing etcd data in: /tmp/integration_test_etcd_data3639025844
{"level":"warn","ts":"2022-09-19T11:26:40.302+0200","caller":"auth/store.go:1220","msg":"simple token is not cryptographically signed"}
I0919 11:26:41.655918 3902624 serving.go:342] Generated self-signed cert (/tmp/test-integration-TestTaintBasedEvictions_Taint_based_evictions_for_NodeNotReady_and_200_tolerationseconds3406213547/apiserver.crt, /tmp/test-integration-TestTaintBasedEvictions_Taint_based_evictions_for_NodeNotReady_and_200_tolerationseconds3406213547/apiserver.key)
I0919 11:26:41.655938 3902624 server.go:554] external host was not specified, using 192.168.1.150
I0919 11:26:41.898161 3902624 shared_informer.go:273] Waiting for caches to sync for node_authorizer
I0919 11:26:41.899379 3902624 plugins.go:158] Loaded 7 mutating admission controller(s) successfully in the following order: NamespaceLifecycle,LimitRanger,DefaultTolerationSeconds,DefaultStorageClass,RuntimeClass,DefaultIngressClass,MutatingAdmissionWebhook.
I0919 11:26:41.899390 3902624 plugins.go:161] Loaded 9 validating admission controller(s) successfully in the following order: LimitRanger,PodSecurity,PersistentVolumeClaimResize,RuntimeClass,CertificateApproval,CertificateSigning,CertificateSubjectRestriction,ValidatingAdmissionWebhook,ResourceQuota.
I0919 11:26:41.899730 3902624 instance.go:261] Using reconciler: lease
I0919 11:26:41.995047 3902624 instance.go:574] API group "internal.apiserver.k8s.io" is not enabled, skipping.
W0919 11:26:42.113797 3902624 genericapiserver.go:656] Skipping API authentication.k8s.io/v1beta1 because it has no resources.
W0919 11:26:42.113812 3902624 genericapiserver.go:656] Skipping API authentication.k8s.io/v1alpha1 because it has no resources.
W0919 11:26:42.115253 3902624 genericapiserver.go:656] Skipping API authorization.k8s.io/v1beta1 because it has no resources.
W0919 11:26:42.119641 3902624 genericapiserver.go:6
…
2m35s: 71 runs so far, 8 failures (11.27%)
```
